### PR TITLE
feat: add redis and redis-cluster in limit-req

### PR DESCRIPTION
**Pull request by theweakgod**
### Description

Fixes https://github.com/apache/apisix/issues/10501 (https://github.com/apache/apisix/issues/10501)
Add options for redis and redis-cluster in the limit-req plugin, and use redis as the counter.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by shreemaan-abhishek** [comment]
I would suggest working on this only after redis support for `limit-conn` is merged.

**Comment by theweakgod** [comment]
> I would suggest working on this only after redis support for `limit-conn` is merged.

Yeah, I think so.

**Comment by theweakgod** [comment]
> @theweakgod we may start working on this now. Please make changes carefully and diligently considering the code review we did for `limit-conn`. Your cooperation is highly appreciated <3

ok

**Comment by theweakgod** [comment]
@shreemaan-abhishek @monkeyDluffy6017 I'm ready for the review.

File: apisix/plugins/limit-req.lua [hunk 1]
```diff
@@ -68,8 +104,23 @@ end
 
 
 local function create_limit_obj(conf)
-    core.log.info("create new limit-req plugin instance")
-    return limit_req_new("plugin-limit-req", conf.rate, conf.burst)
+    if conf.policy == "local" then
+        core.log.info("create new limit-req plugin instance")
+        return limit_req_new("plugin-limit-req", conf.rate, conf.burst)
+    elseif conf.policy == "redis" then
+
```
Comment (monkeyDluffy6017) [inline apisix/plugins/limit-req.lua:111]: No need blank here

File: apisix/plugins/limit-req.lua [hunk 2]
```diff
@@ -68,8 +104,23 @@ end
...
+        core.log.info("create new limit-req redis plugin instance")
+
```
Comment (monkeyDluffy6017) [inline apisix/plugins/limit-req.lua:113]: ditto

File: apisix/plugins/limit-req.lua (hunk 1 above)
Comment (theweakgod) [reply apisix/plugins/limit-req.lua:111]: https://github.com/apache/apisix/blob/master/CODE_STYLE.md#blank-line
But I noticed multiple 'elseif' in the code_style that require blank lines. Are you sure we should remove these blanks? By the way, I also think it doesn't look good that way.

**Review by monkeyDluffy6017 — APPROVED** [review]

File: apisix/plugins/limit-req.lua (hunk 1 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/limit-req.lua:111]: I think it should like this 
```
    if conf.policy == "local" then
        core.log.info("create new limit-req plugin instance")
        return limit_req_new("plugin-limit-req", conf.rate, conf.burst)

    elseif conf.policy == "redis" then
        core.log.info("create new limit-req redis plugin instance")
        return redis_single_new("plugin-limit-req", conf, conf.rate, conf.burst)

    elseif conf.policy == "redis-cluster" then
        core.log.info("create new limit-req redis-cluster plugin instance")
        return redis_cluster_new("plugin-limit-req", conf, conf.rate, conf.burst)
    
```

Comment (theweakgod) [reply apisix/plugins/limit-req.lua:111]: yeah, u are right.

**Review by monkeyDluffy6017 — DISMISSED** [review]

File: docs/en/latest/plugins/limit-req.md [hunk 3]
```diff
@@ -33,16 +33,30 @@ The `limit-req` Plugin limits the number of requests to your service using the [
...
+|--------------------------|---------| -------- |-------------|-----------------------------------| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| conn                     | integer | True     |             | conn > 0                          | Maximum number of concurrent requests allowed. Requests exceeding this ratio (and below `conn` + `burst`) will be delayed (configured by `default_conn_delay`).                                                                                                                                                                                                                                       |
```
Comment (shreemaan-abhishek) [inline docs/en/latest/plugins/limit-req.md:38]: conn?

File: docs/en/latest/plugins/limit-req.md [hunk 4]
```diff
@@ -33,16 +33,30 @@ The `limit-req` Plugin limits the number of requests to your service using the [
...
+| burst                    | integer | True     |             | burst >= 0                        | Number of additional concurrent requests allowed to be delayed per second. If the number exceeds this hard limit, they will get rejected immediately.                                                                                                                                                                                                                                                 |
```
Comment (shreemaan-abhishek) [inline docs/en/latest/plugins/limit-req.md:39]: please check the docs, it has unrelated changes

**Review by shreemaan-abhishek — CHANGES_REQUESTED** [review]

File: apisix/plugins/limit-req/limit-req-redis-cluster.lua [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+function _M.new(plugin_name, conf, rate, burst)
+    local red_cli, err = redis_cluster.new(conf, "plugin-limit-req-redis-cluster-slot-lock")
+    if not red_cli then
+        return nil, err
+    end
+    local self = {
+        conf = conf,
+        plugin_name = plugin_name,
+        burst = burst * 1000,
```
[the remark this answers is not in the document]
Comment (theweakgod) [reply apisix/plugins/limit-req/limit-req-redis-cluster.lua:37]: Reference source code: https://github.com/openresty/lua-resty-limit-traffic/blob/master/lib/resty/limit/req.lua#L60C16-L61. This consistency is required here to ensure compatibility between Redis, Redis Cluster, and resty.limit.req.

File: docs/en/latest/plugins/limit-req.md (hunk 4 above)
Comment (theweakgod) [reply docs/en/latest/plugins/limit-req.md]: done

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by shreemaan-abhishek — APPROVED** [review]
